crates/grafbase-sdk/src/extension/resolver.rs

```diff
@@ -107,3 +51,48 @@ pub trait Resolver: Extension {
         definition: FieldDefinition,
     ) -> Result<Box<dyn Subscription>, Error>;
 }
+
+/// A trait for consuming field outputs from streams.
+///
+/// This trait provides an abstraction over different implementations
+/// of subscriptions to field output streams. Implementors should handle
+/// the details of their specific transport mechanism while providing a
+/// consistent interface for consumers.
+pub trait Subscription {
+    /// Retrieves the next field output from the subscription.
+    ///
+    /// Returns:
+    /// - `Ok(Some(FieldOutput))` if a field output was available
+    /// - `Ok(None)` if the subscription has ended normally
+    /// - `Err(Error)` if an error occurred while retrieving the next field output
+    fn next(&mut self) -> Result<Option<FieldOutput>, Error>;
+}
+
+#[doc(hidden)]
+pub fn register<T: Resolver>() {
+    pub(super) struct Proxy<T: Resolver>(T);
+
+    impl<T: Resolver> AnyExtension for Proxy<T> {
+        fn resolve_field(
+            &mut self,
+            context: SharedContext,
+            directive: Directive,
+            definition: FieldDefinition,
+            inputs: FieldInputs,
+        ) -> Result<FieldOutput, Error> {
+            Resolver::resolve_field(&mut self.0, context, directive, definition, inputs)
+        }
+        fn resolve_subscription(
+            &mut self,
+            context: SharedContext,
+            directive: Directive,
+            definition: FieldDefinition,
+        ) -> Result<Box<dyn Subscription>, Error> {
+            Resolver::resolve_subscription(&mut self.0, context, directive, definition)
+        }
+    }
+    crate::component::register_extension(Box::new(|schema_directives, config| {
+        <T as Extension>::new(schema_directives, config)
+            .map(|extension| Box::new(Proxy(extension)) as Box<dyn AnyExtension>)
+    }))
+}
```

Comment:
what's the point of these?

Reply:
I can't do `impl<T: Resolver> AnyExtension for T` because one can implement both `Resolver` and `Authenticator`, so have no choice but to create a proxy type that implements `AnyExtension`